python/ngen_cal/src/ngen/cal/ngen_hooks/ngen_output.py

```diff
@@ -1,95 +1,298 @@
+    from ngen.cal.model import ModelExec
+    from ngen.config.realization import NgenRealization
 
-from typing import TYPE_CHECKING
 
-if TYPE_CHECKING:
-    from ngen.cal.meta import JobMeta
+class _NgenCalModelOutputFn(typing.Protocol):
+    def __call__(self, id: str) -> pd.Series: ...
 
-class TrouteOutput:
 
+@typing.final
+class TrouteOutput:
     def __init__(self, filepath: Path) -> None:
         self._output_file = filepath
-        self._type = filepath.suffix
+        self._ngen_realization: NgenRealization | None = None
 
-    def _get_dataframe(self) -> DataFrame | None:
-        """Get the t-route output raw dataframe from either csv or hdf5 files
+    def ngen_cal_model_configure(self, config: ModelExec) -> None:
+        # avoid circular import
+        from ngen.cal.ngen import NgenBase
 
-        Returns:
-            DataFrame: _description_
-        """
-        if self._type == '.csv':
-            df = pd.read_csv(self._output_file, index_col=0)
-            df.index = df.index.map(lambda x: 'wb-'+str(x))
-            tuples = [ eval(x) for x in df.columns ]
-            df.columns = pd.MultiIndex.from_tuples(tuples)
-        elif self._type == '.hdf5':
-            df = pd.read_hdf(self._output_file)
-            df.index = df.index.map(lambda x: 'wb-'+str(x))
-            df.columns = pd.MultiIndex.from_tuples(df.columns)
-        else:
-            df = None
-        return df
+        assert isinstance(config, NgenBase)
+        assert config.ngen_realization is not None
+        self._ngen_realization = config.ngen_realization
 
     # Try external provided output hooks, if those fail, try this one
     # this will only execute if all other hooks return None (or they don't exist)
     @hookimpl(specname="ngen_cal_model_output", trylast=True)
-    def get_output(self, id: str) -> Series:
-        try:
-            #look for routed data
-            #read the routed flow at the given id
-            df = self._get_dataframe()
-
-            df = df.loc[id]
-            output = df.xs('q', level=1, drop_level=False)
-            #This is a hacky way to get the time index...pass the time around???
-            tnx_file = list(Path(self._output_file).parent.glob("nex*.csv"))[0]
-            tnx_df = pd.read_csv(tnx_file, index_col=0, parse_dates=[1], names=['ts', 'time', 'Q']).set_index('time')
-            dt_range = pd.date_range(tnx_df.index[0], tnx_df.index[-1], len(output.index)).round('min')
-            output.index = dt_range
-            #this may not be strictly nessicary...I think the _evalutate will align these...
-            output = output.resample('1h').first()
-            output.name="sim_flow"
-            return output
-        except FileNotFoundError:
-            print(f"{self._output_file} not found. Current working directory is {Path.cwd()}")
+    def get_output(self, id: str) -> pd.Series | None:
+        assert self._ngen_realization is not None, "ngen realization required"
```

Comment:
You could certainly see it as redundant but I see it more as enforcing / plainly saying the invariants of the method. The cost benefit of the price paid for the check vs. what you visually get when reading the code / the improved static analysis feedback you get from a tool like `mypy` or some other linter IMO far out weights having a redundant check (plus there is always `python -O` if you want to remove `asserts` for performance reasons).

I also added this as a testing precaution. Ive not written a testing harness for the plugin system yet, so testing plugins without enforcing invariants at the `hook` level can easily lead to false positives in unit tests. I thought this was the easiest way to get the best of both worlds until a harness is added.

Reply:
Not concerned about having the redundant assert, was just thinking through how it could possible be triggered and particularly if there was a better error message/description we use to indicate how it got into that bad state

Reply:
The main way in my mind that this could happen is in a unit test that doesn't call `ngen_cal_model_configure` / pass a valid `ngen.config.NgenRealization` instance.

Reply:
Broader context, i've picked up an appreciation for assertions having reviewed some of [tiger beetle's codebase](https://github.com/tigerbeetle/tigerbeetle) / [style guide](https://github.com/tigerbeetle/tigerbeetle/blob/main/docs/TIGER_STYLE.md#safety).

> **Assert all function arguments and return values, pre/postconditions and invariants.** A
    function must not operate blindly on data it has not checked. The purpose of a function is to
    increase the probability that a program is correct. Assertions within a function are part of how
    functions serve this purpose. The assertion density of the code must average a minimum of two
    assertions per function.